### ISSUE TITLE
bug in nansum with non-float64 dtypes

### ISSUE THREAD

**Issue by jreback** [numpy/numpy#6209]
These non-float ``nansum's`` are a little suspect
xref https://github.com/pydata/pandas/pull/10815

```
In [1]: np.__version__
Out[1]: '1.11.0.dev0+59bbc06'

In [2]: np.nansum(np.array([np.nan]))
Out[2]: 0.0

In [3]: np.nansum(np.array([np.nan],dtype='object'))
Out[3]: nan

In [4]: np.nansum(np.array([np.timedelta64('nat')]))
Out[4]: numpy.timedelta64('NaT')
```

**Comment by charris** [comment]
Maybe, although there are  no nans in the last two examples. Nansum only looks for nans in floating types as those are the only types that have nans.

**Comment by njsmith** [comment]
Yeah, I'm not sure what to make of the object dtype case either -- it's not
like python provides an __isnan__ predicate for generic objects. (Though
OTOH I guess special casing float and numpy scalars wouldn't be that bad,
because it's such a restricted case -- if that's what the weirdos calling
nansum on object arrays want, then maybe there's no harm in giving it to
them ;-).) Handling NaT as a NaN seems like it'd be pretty uncontroversial,
though? NaT is pretty explicitly "this is NaN, but for times".

**Comment by jreback** [comment]
just thinking for consistency, a bit unexpected for the function to do different things depending on the dtype

**Comment by jreback** [comment]
can you put a pointer to the original motivation for this change.

**Comment by seberg** [comment]
@jreback you mean the change that nansum will not return NaN? I think we decided it seemed more right when chuck did some cleanup of the NaN stuff, there is actually a FutureWarning in place for it in NumPy 1.8.
The motivation is that for `np.nansum` NaN is simply ignored, which leaves `[]`, and the sum of that is well defined to be 0.

**Comment by jreback** [comment]
@seberg ok i c. 

was their an original issue / discussion?

**Comment by seberg** [comment]
A bit hard to find, I think it can be traced back to this discussion (just a random mail from it, it is quite long): http://mail.scipy.org/pipermail/numpy-discussion/2013-July/067160.html

**Comment by kjordahl** [comment]
Previous discussion on related issue: #1721.

**Comment by charris** [comment]
@jreback Want to keep this open?

**Comment by jreback** [comment]
you guys had lots of discussion in #1721 about the behavior of ``nansum``, so I won't reargue that (though I don't agree with the approach :)

but assuming that you like the behavior for floats. I certainly think this should be consistent across dtypes (that are summable), e.g. ``np.timedelta64``. of course ``int64`` doesn't have nans (and ``datetime64`` doesn't make sense to sum). ``object`` I think should also be consistent, but I can understand its 'different'.
